src/main/java/com/google/devtools/build/lib/skyframe/PackageLookupFunction.java

```diff
@@ -313,7 +313,7 @@ private PackageLookupValue getPackageLookupValue(
   private static boolean isPackageIgnored(
       PackageIdentifier id, IgnoredPackagePrefixesValue ignoredPatternsValue) {
     PathFragment packageFragment = id.getPackageFragment();
-    for (PathFragment pattern : ignoredPatternsValue.getPatterns()) {
+    for (PathFragment pattern : ignoredPatternsValue.getPatterns(packageFragment)) {
```

Comment:
Thanks for taking the time to review!
From my understanding, `Bazel` pass the set of ignored directories around different components in some case. In this case they consume it immediately. Just checking the path fragment against starting with any of the ignored paths and matching any of the globs only work in this application but not all. I place the logic here because `PackageLookupFunction` seems to be the first place to check ignored directories from `.bazelignore`.

For example, we have a directory structure like the following:
```
/package
├── dir1
└── dir2
    ├── subDir1
    └── subDir2
```
and our `.bazelignore` is:
```
**/subDir*
```
Then only checking is the pathFragment (in this case `/root`) ignored will be not ignore `subDir1` and `subDir2`, we have to recursively read into the directories to check `subDir1` and `subDir2` against the glob.

I also place the logic in SkyValue instead of SkyFunction because `IgnoredPackagePrefixesFunction` will not re-evaluate ignored prefixes when new directories are added, which cause them not being ignored even they match a wildcard pattern in `.bazelignore`.

For example:
```
in .bazelignore:
a/c
a/d
---------------------
WORKSPACE
a/b/BUILD
a/c/BUILD

bzl query //... # should only have targets in b/BUILD

add a/d/BUILD
bzl query //... 
```
If we place the logic in `IgnoredPackagePrefixesFunction`, `a/d/BUILD` will not be ignored because `IgnoredPackagePrefixesFunction` won't recompute.

Do I answer the question or understand it wrong? Thanks again for reviewing.

Reply:
My understanding is that my suggestion would address the code you modified. I agree that there are other parts of the code that would also need to be changed, but I actually think that you could use my suggestion for all of them, suitably modified.

In the case you give, my suggestion will work fine: only when subDir1 and subDir2 are visited will they be excluded. I don't see what the problem is with recursively reading into the directories to check subDir1 and subDir2. Your code also reads recursively into those directories, but it does so eagerly, even if subDir1 and subDir2 are never used in the build.

You won't need the mutable SkyValue if you don't do a directory listing, which is my suggestion.

Just to be explicit, the PR in its current form is not viable. The uncached system calls made on what should be a quick lookup, coupled with the mutable SkyValue, make this a no-go. Please try implementing my suggestion and creating a new PR. If it doesn't meet your needs, add a test case demonstrating the failure and we can discuss it there.

Reply:
Hi @janakdr! I tried implement it in the way you suggested (https://github.com/bazelbuild/bazel/pull/13807) but the wildcard test cases will failed. Could you please take a look and make sure I am on the right track? Thanks!